### PR TITLE
fix(masterchef) ubq rewards

### PR DESCRIPTION
**Pull request by zapaz**
Fix MasterChef2 UBQ rewards calculation

> we had the amount of bonding shares that account for your "shares" (aka right to mint UBQ) and now we have only 1 Bonding Share and link to that BS the amount of "shares" . The problem is that its still divided by the supply of BS instead of just calling _totalShares
> 
>  add existing tests to this pr to have an extended test coverage
> 
> and will have to redeploy the masterchefV2 SC
> before that, need to remove all the roles from the previous one
> 
> the LP amount /shares is now linked to the BondingShareV2 so there is no need to redeploy something else 
> but some tests needed to make sure of that
> 
> and include the audit required modification before redeploying...
> 
> zgorizzo69

### CODE REVIEW

**Comment by zapaz** [comment]
**DEPLOYMENT MasterChefV2.1 PROCESS**

1/ Deploy MasterChefV2.1 SmartContract to mainnet

2/ Tests : 
- Check initialization with 5 deposits is ok
- Check simulateMigrate on new deployment (after local fork)
- Other manual checks

3/ Set UBQ_MINTER_ROLE to new MasterChefV2
4/ Set setUGOVPerBlock to one (1e18)

5/ Set new MasterChefV2 address in manager

File: contracts/MasterChefV2.sol [hunk 1]
```diff
@@ -42,11 +43,11 @@ contract MasterChefV2 {
     UbiquityAlgorithmicDollarManager public manager;
 
     // uGOV tokens created per block.
-    uint256 public uGOVPerBlock = 1e18;
+    uint256 public uGOVPerBlock;
```
Comment (zgorizzo69) [inline contracts/MasterChefV2.sol:46]: if it is not set there it could be setup in the constructor

File: deploy/V2.ts [hunk 2]
```diff
@@ -1,28 +1,132 @@
 import { HardhatRuntimeEnvironment } from "hardhat/types";
```
Comment (zgorizzo69) [inline deploy/V2.ts:1]: we should have left this deployment alone and create a new one just for the masterchef V2.1 update

File: test/SushiSwapPool.test.ts [hunk 3]
```diff
@@ -217,7 +217,7 @@ describe("SushiSwapPool", () => {
       expect(await newSushi.pair()).to.equal(await sushiUGOVPool.pair());
     });
 
-    it("should add pool and earn sushi", async () => {
+    it.skip("should add pool and earn sushi", async () => {
```
Comment (zgorizzo69) [inline test/SushiSwapPool.test.ts:220]: I don't understand why we would skip an old test that was passing ...we can remove the `skip`

File: test/utils/hardhatNode.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
-import { ContractTransaction } from "ethers";
-import { network, ethers } from "hardhat";
+import { ContractTransaction, Signer, BigNumber } from "ethers";
```
Comment (zgorizzo69) [inline test/utils/hardhatNode.ts:1]: ethers should come from hardhat so to use bignumber just write ethers.BigNumber as it is imported on the next line

**Review by zgorizzo69 — CHANGES_REQUESTED** [review]

File: contracts/MasterChefV2.sol (hunk 1 above)
Comment (pavlovcik) [reply contracts/MasterChefV2.sol:46]: we will deploy with `uGOVPerBlock = 0`

**Comment by zgorizzo69** [comment]
> **DEPLOYMENT MasterChefV2.1 PROCESS**
> 
> 1/ Deploy MasterChefV2.1 SmartContract to mainnet with script
> 
> `npx hardhat deploy --tags MasterChefV2.1 --network mainnet`
> 
> _To be launch with private key of UBQ manager_
> 
> 2/ Tests :
> 
>     * Check initialization with 5 deposits is ok
> 
>     * Check simulateMigrate on new deployment (after local fork)
> 
>     * Other manual checks
> 
> 
> 3/ Set manually UBQ_MINTER_ROLE to new MasterChefV2
> 4/ Set manually setUGOVPerBlock to one (1e18)
> 
> 5/ Set manually new MasterChefV2 address in manager
> 
> _manually : can be done via etherscan or other means_

I think we should set the UGOVPerBlock in the constructor to save some gas ;)

File: test/SushiSwapPool.test.ts (hunk 3 above)
Comment (pavlovcik) [reply test/SushiSwapPool.test.ts:220]: > > **DEPLOYMENT MasterChefV2.1 PROCESS**
> > 1/ Deploy MasterChefV2.1 SmartContract to mainnet with script
> > `npx hardhat deploy --tags MasterChefV2.1 --network mainnet`
> > _To be launch with private key of UBQ manager_
> > 2/ Tests :
> > ```
> > * Check initialization with 5 deposits is ok
> > 
> > * Check simulateMigrate on new deployment (after local fork)
> > 
> > * Other manual checks
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 3/ Set manually UBQ_MINTER_ROLE to new MasterChefV2
> > 4/ Set manually setUGOVPerBlock to one (1e18)
> > 5/ Set manually new MasterChefV2 address in manager
> > _manually : can be done via etherscan or other means_
> 
> I think we should set the UGOVPerBlock in the constructor to save some gas ;)

we're here to chew bubble gum and burn some gas...and we're all out of bubble gum

File: test/utils/hardhatNode.ts (hunk 4 above)
Comment (zapaz) [reply test/utils/hardhatNode.ts:1]: yes, will do that

File: test/SushiSwapPool.test.ts (hunk 3 above)
Comment (zapaz) [reply test/SushiSwapPool.test.ts:220]: both estimation gas problem after london hardfork + too long calculation on MasterChefV1 (endless on local computer), on this line : 
`await masterChef.connect(sushiChef).add(totAllocPoint, uGOVPair, true);`

that is Masterchef V1 problem...

**Pull request merged**